src/panels/logbook/ha-logbook.ts

```diff
@@ -254,6 +276,97 @@ class HaLogbook extends LitElement {
     this._savedScrollPos = (e.target as HTMLDivElement).scrollTop;
   }
 
+  private _formatEventBy = (item: LogbookEntry, seenEntities: string[]) => {
+    if (item.context_event_type === "call_service") {
+      return `${this.hass.localize("ui.components.logbook.from_service")} ${
+        item.context_domain
+      }.${item.context_service}`;
+    }
+    if (
+      item.context_event_type === "automation_triggered" &&
+      item.context_entity_id
+    ) {
+      if (seenEntities.includes(item.context_entity_id)) {
+        return ``;
+      }
+      seenEntities.push(item.context_entity_id);
+      return html`${this.hass.localize("ui.components.logbook.from_automation")}
+        <a
+          href="#"
+          @click=${this._entityClicked}
+          .entityId=${item.context_entity_id}
+          class="name"
+          >${item.context_name}</a
+        >`;
+    }
+    if (item.context_name) {
+      return `${this.hass.localize("ui.components.logbook.from")} ${
+        item.context_name
+      }`;
+    }
+    if (item.context_event_type === "state_changed") {
+      return ``;
+    }
+    if (item.context_event_type! in EVENT_LOCALIZE_MAP) {
+      return `${this.hass.localize(
+        `ui.components.logbook.${EVENT_LOCALIZE_MAP[item.context_event_type!]}`
+      )}`;
+    }
+    return `${this.hass.localize(
+      "ui.components.logbook.from"
+    )} ${this.hass.localize("ui.components.logbook.event")} ${
+      item.context_event_type
+    }`;
+  };
+
+  private _formatMessageWithPossibleEntity(
+    message: string,
+    seenEntities: string[],
+    forEntityId?: string,
+    localizePrefix?: string
+  ) {
+    const matches = message.match(/state of ([^ ]+)/);
+    if (matches) {
+      const entityId = matches[1];
+      if (seenEntities.includes(entityId)) {
+        return ``;
+      }
+      seenEntities.push(entityId);
+      const parts = message.split(entityId);
+      return html` ${localizePrefix ? this.hass.localize(localizePrefix) : ""}
+        ${parts[0]}
+        <a
+          href="#"
+          @click=${this._entityClicked}
+          .entityId=${entityId}
+          class="name"
+          >${this.hass.states[entityId].attributes.friendly_name || entityId}</a
+        >
+        ${parts[1]}`;
+    }
+    if (forEntityId) {
+      const forEntityName =
+        this.hass.states[forEntityId].attributes.friendly_name;
+      if (forEntityName && message.endsWith(forEntityName)) {
```

Comment:
This really could be starts or ends with but in practice we don't have ends right now